File: Java/src/com/tankbattle/Bullet.java
package com.tankbattle;

import java.awt.*;
import java.util.ArrayList;
import java.util.Iterator;

import static com.tankbattle.Utils.*;

/**
 * 子弹类（含拖尾效果）
 */
public class Bullet {
    public double x, y;
    public int w, h;
    public Utils.Vec2 dir;
    public double speed;
    public boolean isPlayer;
    public int playerId;
    public Color color;
    public boolean alive;

    // ===== 拖尾 =====
    private static class TrailParticle {
        double x, y;
        double size;
        double age;
        double lifetime = 0.25;  // 拖尾寿命（秒）
    }
    private ArrayList<TrailParticle> trail;

    public Bullet(double x, double y, Utils.Vec2 dir, boolean isPlayer, int playerId, Color color) {
        this.x = x - BULLET_SIZE / 2.0;
        this.y = y - BULLET_SIZE / 2.0;
        this.w = BULLET_SIZE;
        this.h = BULLET_SIZE;
        this.dir = dir;
        this.speed = BULLET_SPEED;
        this.isPlayer = isPlayer;
        this.playerId = playerId;
        this.color = color;
        this.alive = true;
        this.trail = new ArrayList<>();
    }

    public Rectangle getRect() {
        return new Rectangle((int) x, (int) y, w, h);
    }

    public void update(double dt) {
        // 添加拖尾粒子
        TrailParticle p = new TrailParticle();
        p.x = this.x + this.w / 2.0;
        p.y = this.y + this.h / 2.0;
        p.size = this.w * 0.6;
        p.lifetime = 0.15;
        p.age = 0;
        trail.add(p);

        // 更新位置
        x += dir.x * speed * dt;
        y += dir.y * speed * dt;

        // 更新拖尾
        Iterator<TrailParticle> it = trail.iterator();
        while (it.hasNext()) {
            TrailParticle tp = it.next();
            tp.age += dt;
            tp.size *= (1.0 - dt / tp.lifetime);
            if (tp.age >= tp.lifetime || tp.size < 0.5) {
                it.remove();
            }
        }

        if (x < -20 || x > WINDOW_WIDTH + 20 || y < -20 || y > WINDOW_HEIGHT + 20) {
            alive = false;
        }
    }

    public void draw(Graphics2D g) {
        // ===== 画拖尾（用同步块 + 复制） =====
        ArrayList<TrailParticle> trailCopy;
        synchronized (trail) {
            trailCopy = new ArrayList<>(trail);
        }
    
        for (TrailParticle tp : trailCopy) {
            double alpha = 1.0 - tp.age / tp.lifetime;
            int size = (int) Math.round(tp.size);
            if (size > 0 && alpha > 0.05) {
                int a = (int) (alpha * 200);
                Color col = new Color(
                    color.getRed(),
                    color.getGreen(),
                    color.getBlue(),
                    Math.min(a, 255)
                );
                g.setColor(col);
                g.fillOval(
                    (int) (tp.x - size),
                    (int) (tp.y - size),
                    size * 2,
                    size * 2
                );
            }
        }

        // ===== 画子弹 =====
        if (!alive) return;
        g.setColor(color);
        int cx = (int) (x + w / 2.0);
        int cy = (int) (y + h / 2.0);
        g.fillOval(cx - w / 2, cy - h / 2, w, h);
        g.setColor(Color.WHITE);
        g.fillOval(cx - w / 4, cy - h / 4, w / 2, h / 2);
    }
}

File: C++/src/Bullet.cpp
#include "Bullet.h"
#include "resource/TextureManager.h"

Bullet::Bullet(float x, float y, float dx, float dy, bool isPlayer, int pid, sf::Color col)
    : x(x - getBulletSize()/2.0f), y(y - getBulletSize()/2.0f),
      w(getBulletSize()), h(getBulletSize()),
      dx(dx), dy(dy), speed(getBulletSpeed()),
      player(isPlayer), playerId(pid), color(col), alive(true), damage(1) {}

void Bullet::update(float dt) {
    // 添加拖尾粒子
    TrailParticle p;
    p.x = x + w/2;
    p.y = y + h/2;
    p.size = w * 0.6f;
    p.age = 0;
    p.lifetime = 0.15f;
    trail.push_back(p);

    // 更新位置
    x += dx * speed * dt;
    y += dy * speed * dt;

    // 更新拖尾
    for (auto it = trail.begin(); it != trail.end(); ) {
        it->age += dt;
        it->size *= (1.0f - dt / it->lifetime);
        if (it->age >= it->lifetime || it->size < 0.5f) {
            it = trail.erase(it);
        } else {
            ++it;
        }
    }

    if (x < -20 || x > WINDOW_WIDTH + 20 || y < -20 || y > WINDOW_HEIGHT + 20) {
        alive = false;
    }
}

void Bullet::draw(sf::RenderWindow& window) {
    // 画拖尾
    for (const auto& p : trail) {
        float alpha = 1.0f - p.age / p.lifetime;
        int size = static_cast<int>(p.size);
        if (size > 0 && alpha > 0.05f) {
            sf::CircleShape shape(size);
            shape.setPosition(sf::Vector2f(p.x - size, p.y - size));
            sf::Color col = color;
            col.a = static_cast<uint8_t>(alpha * 255);
            shape.setFillColor(col);
            window.draw(shape);
        }
    }

    if (!alive) return;

    // ===== 尝试获取贴图 =====
    std::string entityId;
    if (player) {
        entityId = "bullet_p" + std::to_string(playerId);
    } else {
        entityId = "bullet_enemy";
    }

    auto& tm = TextureManager::getInstance();
    auto texture = tm.getEntityTexture(entityId);

    int cx = static_cast<int>(x + w / 2.0f);
    int cy = static_cast<int>(y + h / 2.0f);

    if (texture) {
        // ===== 使用贴图绘制 =====
        sf::Sprite sprite(*texture);
        sprite.setPosition(sf::Vector2f(cx - w/2.0f, cy - h/2.0f));
        sprite.setScale(sf::Vector2f(
            w / (float)texture->getSize().x,
            h / (float)texture->getSize().y
        ));
        window.draw(sprite);
        return;
    }

    sf::CircleShape shape(w / 2.0f);
    shape.setPosition(sf::Vector2f(x, y));
    shape.setFillColor(color);
    shape.setOutlineColor(sf::Color::White);
    shape.setOutlineThickness(1.0f);
    window.draw(shape);
}

sf::FloatRect Bullet::getRect() const {
    return sf::FloatRect(sf::Vector2f(x, y), sf::Vector2f(w, h));
}